### PR TITLE
Use local identities to represent CIDR

### PR DESCRIPTION
**Pull request by tgraf**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6465)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/identity/numericidentity.go [hunk 1]
```diff
@@ -342,3 +346,8 @@ func IterateReservedIdentities(f func(key string, value NumericIdentity)) {
 		f(key, value)
 	}
 }
+
+// HasLocalScope returns true if the identity has a local scope
+func (id NumericIdentity) HasLocalScope() bool {
```
Comment (ianvernon) [inline pkg/identity/numericidentity.go:377]: Does the allocator for global identities have to be made aware that this bit is 'reserved', so to speak, for indicating whether the identity is local vs. global? in other words, how do we ensure that the allocator for global identities does not conflict with the allocator for local CIDR identities?

File: pkg/identity/cache/allocation_test.go [hunk 2]
```diff
@@ -177,3 +177,39 @@ func (ias *IdentityAllocatorSuite) TestAllocator(c *C) {
...
+
+	InitIdentityAllocator(dummyOwner{})
+	defer Close()
+	defer IdentityAllocator.DeleteAllKeys()
+
+	id, isNew, err := AllocateIdentity(lbls1)
+	c.Assert(id, Not(IsNil))
+	c.Assert(err, IsNil)
+	c.Assert(isNew, Equals, true)
+	c.Assert(id.ID&identity.LocalIdentityFlag != 0, Equals, true)
```
Comment (ianvernon) [inline pkg/identity/cache/allocation_test.go]: why not `id.ID. HasLocalScope()`?

File: pkg/identity/cache/allocation_test.go [hunk 3]
```diff
@@ -177,3 +177,39 @@ func (ias *IdentityAllocatorSuite) TestAllocator(c *C) {
...
+	c.Assert(Release(id), IsNil)
+
+	cache = GetIdentityCache()
+	c.Assert(cache[id.ID], IsNil)
+
+	id, isNew, err = AllocateIdentity(lbls1)
+	c.Assert(id, Not(IsNil))
+	c.Assert(err, IsNil)
+	c.Assert(isNew, Equals, true)
+	c.Assert(id.ID&identity.LocalIdentityFlag != 0, Equals, true)
```
Comment (ianvernon) [inline pkg/identity/cache/allocation_test.go]: Same here.

File: pkg/identity/cache/local.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+// lookup searches for a local identity matching the given labels and returns
+// it. If found, the reference count is NOT incremented and thus release must
+// NOT be called.
+func (l *localIdentityCache) lookup(lbls labels.Labels) *identity.Identity {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	stringRepresentation := string(lbls.SortedList())
```
Comment (ianvernon) [inline pkg/identity/cache/local.go:154]: This may be outside the scope of this PR, but we sort labels a lot in these functions. Is there any way to ensure that the labels are sorted further up the call-stack? This could prove to be an unreasonable expectation on my side.

File: pkg/identity/cache/local.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+func (l *localIdentityCache) bumpNextNumericIdentity() {
+	if l.nextNumericIdentity == l.maxID {
+		l.nextNumericIdentity = l.minID
+	} else {
+		l.nextNumericIdentity++
+	}
+}
+
+// allocateNumericIdentity returns the next available numeric identity or an error
+func (l *localIdentityCache) allocateNumericIdentity() (identity.NumericIdentity, error) {
```
Comment (jrajahalme) [inline pkg/identity/cache/local.go]: IMO this would be better called `getNextFreeNumericIdentity()` as this alone does not mark the identity as taken/allocated.

File: pkg/ipcache/ipcache.go [hunk 6]
```diff
@@ -48,6 +48,10 @@ const (
 	// FromAgentLocal is the source used for identities derived during the
 	// agent bootup process. This includes identities for endpoint IPs.
 	FromAgentLocal Source = "agent-local"
+
+	// FromCIDR is the source used for identities that have been derived
+	// from local CIDR representations
+	FromCIDR = "cidr"
```
Comment (jrajahalme) [inline pkg/ipcache/ipcache.go]: Would this benefit from typing this as `Source`?

File: pkg/ipcache/ipcache.go [hunk 7]
```diff
@@ -350,6 +356,11 @@ func (ipc *IPCache) deleteLocked(ip string) {
 		return
 	}
 
+	if cachedIdentity.Source != source {
+		scopedLog.Debugf("Skipping delete of identity from source %s", source)
+		return
+	}
+
```
Comment (jrajahalme) [inline pkg/ipcache/ipcache.go]: Would it ease backporting if the bug fix would be separated to it's own commit?

File: pkg/identity/cache/local.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+	id := &identity.Identity{
+		ID:             numericIdentity | identity.LocalIdentityFlag,
+		Labels:         lbls,
+		LabelArray:     lbls.LabelArray(),
+		ReferenceCount: 1,
+	}
+
+	l.identitiesByLabels[stringRepresentation] = id
+	l.identitiesByID[numericIdentity] = id
```
Comment (jrajahalme) [inline pkg/identity/cache/local.go:102]: IMO it would be clearer and less error-prone to have the map key include the local identity flag. Maybe it should already be returned by the "allocator"?

**Review by jrajahalme — APPROVED** [review]
But small nits..

File: pkg/identity/numericidentity.go [hunk 9]
```diff
@@ -29,6 +29,10 @@ const (
 	// shifted
 	ClusterIDShift = 16
 
+	// LocalIdentityFlag is the bit in the numeric identity that identifies
+	// a numeric identity to have local scope
+	LocalIdentityFlag = NumericIdentity(1 << 24)
+
```
Comment (raybejjani) [inline pkg/identity/numericidentity.go:35]: I can't comment on it directly, but the comment on  `type NumericIdentity` also needs an (line 288)

File: pkg/identity/cache/local.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+	firstID := l.nextNumericIdentity
+	for {
+		if _, ok := l.identitiesByID[l.nextNumericIdentity]; !ok {
+			id := l.nextNumericIdentity
+			l.bumpNextNumericIdentity()
+			return id, nil
+		}
+
+		l.bumpNextNumericIdentity()
+		if l.nextNumericIdentity == firstID {
```
Comment (nebril) [inline pkg/identity/cache/local.go:68]: Could `l.nextNumericIdentity` be compared to `l.minID` to indicate that there was an overflow? Or do we want to iterate over all possible identities (in which case this is fine).

File: pkg/identity/cache/local.go (hunk 5 above)
Comment (nebril) [inline pkg/identity/cache/local.go]: Can you add comment that this method should be used while holding the lock?

File: pkg/identity/cache/local.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+		case id.ReferenceCount > 1:
+			id.ReferenceCount--
+			return false
+
+		case id.ReferenceCount == 1:
+			// Release is only attempted once, when the reference count is
+			// hitting the last use
+			stringRepresentation := string(id.Labels.SortedList())
+			delete(l.identitiesByLabels, stringRepresentation)
+			delete(l.identitiesByID, (id.ID & ^identity.LocalIdentityFlag))
```
Comment (raybejjani) [inline pkg/identity/cache/local.go]: Will making the map key this way be a problem later? If we ever use bits higher than 24? should this be something like ` (id.ID & (identity.LocalIdentityFlag-1))` to only allow the lower 23 bits (assuming it is a simple bit shift to make LocalIdentityFlag)

File: pkg/identity/cache/allocator.go [hunk 12]
```diff
@@ -155,15 +160,21 @@ func AllocateIdentity(lbls labels.Labels) (*identity.Identity, bool, error) {
...
+				logfields.Identity:       reservedIdentity.ID,
+				logfields.IdentityLabels: lbls.String(),
+				"isNew":                  false,
+			}).Debug("Resolved reserved identity")
+			return reservedIdentity, false, nil
+		}
+
+		if localIdentities != nil {
+			return localIdentities.lookupOrCreate(lbls)
+		}
```
Comment (raybejjani) [inline pkg/identity/cache/allocator.go]: Falling through to a global identity when `localIdentities != nil` is intended right? This way there is a fallback if we make a mistake with `RequiresGlobalIdentity`?

File: pkg/identity/cache/allocator.go [hunk 13]
```diff
@@ -126,6 +130,7 @@ func Close() {
 	IdentityAllocator.Delete()
 	watcher.stop()
 	IdentityAllocator = nil
+	localIdentities = nil
```
Comment (raybejjani) [inline pkg/identity/cache/allocator.go:133]: It might be good to mention in the function comment that calling `InitIdentityAllocator` again would reuse the local identities (I doubt this will ever really come up, though).

File: pkg/identity/numericidentity.go (hunk 1 above)
Comment (tgraf) [reply pkg/identity/numericidentity.go:377]: The allocator was already limited to not overlap with this flag. I've changed the first commit to move the constants defining this to be all in one place for easier visibility.

File: pkg/identity/cache/local.go (hunk 4 above)
Comment (tgraf) [reply pkg/identity/cache/local.go:154]: I think we can cache this but I agree, outside of the scope of this PR.

File: pkg/ipcache/ipcache.go (hunk 7 above)
Comment (tgraf) [reply pkg/ipcache/ipcache.go]: 👍 Factored out into https://github.com/cilium/cilium/pull/6475

File: pkg/identity/cache/local.go (hunk 10 above)
Comment (tgraf) [reply pkg/identity/cache/local.go:68]: We want to iterate over all possible identities and allow reuse after the initial pass over (which is already highly unlikely) given the total space of 24bits.

**Comment by tgraf** [comment]
test-me-please

**Review by raybejjani — APPROVED** [review]
LGTM :) just comments about comments

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
I reran the tests since the two fqdn fails might simply be small timeouts (3s) and the other test fail was an interrupted test.
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8604/

**Comment by tgraf** [comment]
The FQDN test failure seems legitimate. Will dig in.

**Comment by tgraf** [comment]
Confirming that for FQDN policies are getting correct local identities but the ipcache plumbing is skipped for some reaosn

**Comment by tgraf** [comment]
Found bug in ipcache event invocation. Fixed

**Comment by tgraf** [comment]
test-me-please

File: pkg/identity/cache/allocator.go (hunk 12 above)
Comment (tgraf) [reply pkg/identity/cache/allocator.go]: It's always better to fall back to global identities and provide correct behavior even if it is not using the most efficient implementation.

File: pkg/identity/cache/local.go (hunk 11 above)
Comment (tgraf) [reply pkg/identity/cache/local.go]: Jarno brought this up as well, changed

File: pkg/identity/cache/allocator.go (hunk 13 above)
Comment (tgraf) [reply pkg/identity/cache/allocator.go:133]: Not sure I understand, `InitIdentityAllocator()` will create a new local cache

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

```
00:12:37.364     k8s1-1.13: Get https://quay.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

```

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

k8s tests passed
runtime tests hit panic fixed by #6552

**Comment by tgraf** [comment]
test-me-please
